src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -538,23 +555,54 @@ repl_req_ptr_t RaftReplDev::applier_create_req(repl_key const& rkey, journal_typ
     }
 
     // rreq->init will allocate the block if it has linked data.
-    auto status = rreq->init(rkey, code, false /* is_proposer */, user_header, key, data_size, m_listener);
-    if (!rreq->has_linked_data()) { return rreq; }
-#ifdef _PRERELEASE
-    if (is_data_channel) {
-        if (iomgr_flip::instance()->test_flip("fake_reject_append_data_channel")) {
-            LOGINFO("Data Channel: Reject append_entries flip is triggered for rkey={}", rkey.to_string());
-            status = ReplServiceError::NO_SPACE_LEFT;
-        }
-    } else {
-        if (iomgr_flip::instance()->test_flip("fake_reject_append_raft_channel")) {
-            LOGINFO("Raft Channel: Reject append_entries flip is triggered for rkey={}", rkey.to_string());
-            status = ReplServiceError::NO_SPACE_LEFT;
-        }
-    }
-#endif
+    auto status = init_req_ctx(rreq, rkey, code, false /* is_proposer */, user_header, key, data_size, m_listener);
+
     if (status != ReplServiceError::OK) {
-        RD_LOGD("For Repl_key=[{}] alloc hints returned error={}, failing this req", rkey.to_string(), status);
+        if (status == ReplServiceError::NO_SPACE_LEFT && !is_data_channel && !rreq->is_proposer()) {
+            RD_LOGD("For Repl_key=[{}] alloc hints returned error={} , lsn={}", rkey.to_string(), status, lsn);
+
+            // only when the follower got no_space_left error in raft channel, we try to handle it. if it is a leader,
+            // we will not handle it and return fail to client.
+
+            pause_statemachine();
+            // since we pause statemachine here, handle_no_space_left will either be called in the commit thread, or be
+            // called here, where the raft channel is blocked
+
+            // the chunk_id is the chunk on which no_space_left happens
+            const auto& chunk_id = rreq->local_blkid().chunk_num();
+            const auto last_append_lsn = m_data_journal->next_slot() - 1;
+            const auto last_commit_lsn = get_last_commit_lsn();
+
+            RD_LOGD("no_space_left occurs in raft channel for chunk_id={}, last_append_lsn={}, last_commit_lsn={}",
+                    chunk_id, last_append_lsn, last_commit_lsn);
+
+            if (s_cast< repl_lsn_t >(last_append_lsn) == last_commit_lsn) {
```

Comment:
i  think it should be `lsn == last_commit_lsn` ?

Reply:
if all the appended logs are committed, we can say the state is clean and we can start handling no_space_left.